File: health_prompt/HealthDashboard/Controllers/HDProfileViewController.m
//
//  HDProfileViewController.m
//  HealthDashboard - 个人中心
//
//  Created by zhang, haizi on 2026/3/19.
//

#import "HDProfileViewController.h"
#import "../Models/HDHealthDataModel.h"

@interface HDProfileViewController ()
@property (nonatomic, strong) UIScrollView *scrollView;
@property (nonatomic, strong) UIView       *contentView;
@property (nonatomic, strong) UISwitch     *themeSwitch;
@end

@implementation HDProfileViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setupScrollView];
    [self buildContent];
    [self applyTheme];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    self.navigationController.navigationBarHidden = YES;
    [self applyTheme];
}

- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];
    _scrollView.frame = self.view.bounds;
}

- (void)setupScrollView {
    CGFloat sw = UIScreen.mainScreen.bounds.size.width;
    CGFloat sh = UIScreen.mainScreen.bounds.size.height;
    _scrollView = [[UIScrollView alloc] initWithFrame:CGRectMake(0,0,sw,sh)];
    _scrollView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    _scrollView.showsVerticalScrollIndicator = NO;
    [self.view addSubview:_scrollView];
    _contentView = [[UIView alloc] initWithFrame:CGRectMake(0,0,sw,600)];
    [_scrollView addSubview:_contentView];
    _scrollView.contentSize = CGSizeMake(sw, 600);
}

- (void)buildContent {
    CGFloat sw = self.view.bounds.size.width;
    CGFloat topY = 60;

    // 页面标题
    UILabel *title = [[UILabel alloc] initWithFrame:CGRectMake(20, topY, sw-40, 40)];
    title.tag  = 801;
    title.text = @"个人中心";
    title.font = [UIFont systemFontOfSize:28 weight:UIFontWeightHeavy];
    title.textColor = UIColor.whiteColor;
    [_contentView addSubview:title];

    // 头像
    UIView *avatar = [[UIView alloc] initWithFrame:CGRectMake(sw/2-44, topY+58, 88, 88)];
    avatar.backgroundColor    = [UIColor colorWithRed:0.22 green:0.54 blue:0.95 alpha:1];
    avatar.layer.cornerRadius = 44;
    [_contentView addSubview:avatar];
    UILabel *avatarEmoji = [[UILabel alloc] initWithFrame:avatar.bounds];
    avatarEmoji.text = @"😊";
    avatarEmoji.font = [UIFont systemFontOfSize:44];
    avatarEmoji.textAlignment = NSTextAlignmentCenter;
    [avatar addSubview:avatarEmoji];

    // 用户名
    UILabel *name = [[UILabel alloc] initWithFrame:CGRectMake(20, topY+158, sw-40, 28)];
    name.tag  = 802;
    name.text = @"健康达人";
    name.font = [UIFont systemFontOfSize:20 weight:UIFontWeightBold];
    name.textColor     = UIColor.whiteColor;
    name.textAlignment = NSTextAlignmentCenter;
    [_contentView addSubview:name];

    // 统计行
    HDHealthDataModel *m = [HDHealthDataModel shared];
    NSArray *statTitles = @[@"今日步数", @"喝水", @"睡眠"];
    NSArray *statValues = @[
        [NSString stringWithFormat:@"%ld", (long)m.todaySteps],
        [NSString stringWithFormat:@"%.0fml", m.waterML],
        [NSString stringWithFormat:@"%.1fh", [m.sleepHours.lastObject floatValue]]
    ];
    CGFloat itemW = sw / 3.0;
    CGFloat statsY = topY + 200;
    for (NSInteger i = 0; i < 3; i++) {
        UILabel *val = [[UILabel alloc] initWithFrame:CGRectMake(i*itemW, statsY, itemW, 32)];
        val.text = statValues[i];
        val.font = [UIFont monospacedDigitSystemFontOfSize:20 weight:UIFontWeightBold];
        val.textColor = [UIColor colorWithRed:0.22 green:0.54 blue:0.95 alpha:1];
        val.textAlignment = NSTextAlignmentCenter;
        [_contentView addSubview:val];

        UILabel *lbl = [[UILabel alloc] initWithFrame:CGRectMake(i*itemW, statsY+34, itemW, 18)];
        lbl.text = statTitles[i];
        lbl.font = [UIFont systemFontOfSize:12];
        lbl.textColor = [UIColor colorWithWhite:0.55 alpha:1];
        lbl.textAlignment = NSTextAlignmentCenter;
        [_contentView addSubview:lbl];
    }

    // 夜间模式卡片
    CGFloat themeY = statsY + 72;
    UIView *themeCard = [[UIView alloc] initWithFrame:CGRectMake(16, themeY, sw-32, 64)];
    themeCard.tag = 820;
    themeCard.backgroundColor    = [UIColor colorWithRed:0.12 green:0.16 blue:0.24 alpha:1];
    themeCard.layer.cornerRadius = 16;
    [_contentView addSubview:themeCard];

    UILabel *themeLbl = [[UILabel alloc] initWithFrame:CGRectMake(16, 18, sw-100, 28)];
    themeLbl.tag  = 821;
    themeLbl.text = @"🌙  夜间模式";
    themeLbl.font = [UIFont systemFontOfSize:16 weight:UIFontWeightSemibold];
    themeLbl.textColor = UIColor.whiteColor;
    [themeCard addSubview:themeLbl];

    _themeSwitch = [[UISwitch alloc] initWithFrame:CGRectMake(sw-32-70, 16, 51, 31)];
    _themeSwitch.on = m.isDarkMode;
    _themeSwitch.onTintColor = [UIColor colorWithRed:0.22 green:0.54 blue:0.95 alpha:1];
    [_themeSwitch addTarget:self action:@selector(themeSwitchChanged) forControlEvents:UIControlEventValueChanged];
    [themeCard addSubview:_themeSwitch];

    // 目标卡片
    CGFloat goalY = themeY + 80;
    UIView *goalCard = [[UIView alloc] initWithFrame:CGRectMake(16, goalY, sw-32, 120)];
    goalCard.tag = 830;
    goalCard.backgroundColor    = [UIColor colorWithRed:0.12 green:0.16 blue:0.24 alpha:1];
    goalCard.layer.cornerRadius = 16;
    [_contentView addSubview:goalCard];

    UILabel *goalTitle = [[UILabel alloc] initWithFrame:CGRectMake(16, 14, sw-64, 22)];
    goalTitle.tag  = 831;
    goalTitle.text = @"🎯  我的目标";
    goalTitle.font = [UIFont systemFontOfSize:16 weight:UIFontWeightSemibold];
    goalTitle.textColor = UIColor.whiteColor;
    [goalCard addSubview:goalTitle];

    NSArray *goals = @[
        [NSString stringWithFormat:@"每日步数：%ld 步", (long)m.stepsGoal],
        [NSString stringWithFormat:@"每日喝水：%.0f ml", m.waterGoalML]
    ];
    for (NSInteger i = 0; i < 2; i++) {
        UILabel *gl = [[UILabel alloc] initWithFrame:CGRectMake(16, 44 + i*32, sw-64, 24)];
        gl.text = goals[i];
        gl.font = [UIFont systemFontOfSize:14];
        gl.textColor = [UIColor colorWithWhite:0.70 alpha:1];
        [goalCard addSubview:gl];
    }
}

#pragma mark - 主题切换

- (void)themeSwitchChanged {
    [HDHealthDataModel shared].isDarkMode = _themeSwitch.isOn;
    [self applyTheme];
    [[NSNotificationCenter defaultCenter] postNotificationName:@"HDThemeDidChange" object:nil];
}

- (void)applyTheme {
    BOOL dark = [HDHealthDataModel shared].isDarkMode;
    UIColor *bg    = dark ? [UIColor colorWithRed:0.06 green:0.08 blue:0.14 alpha:1.0]
                          : [UIColor colorWithRed:0.93 green:0.95 blue:0.97 alpha:1.0];
    UIColor *cardC = dark ? [UIColor colorWithRed:0.12 green:0.16 blue:0.24 alpha:1.0] : UIColor.whiteColor;
    UIColor *textC = dark ? UIColor.whiteColor : [UIColor colorWithRed:0.10 green:0.12 blue:0.18 alpha:1.0];

    self.view.backgroundColor    = bg;
    _scrollView.backgroundColor  = bg;
    _contentView.backgroundColor = bg;

    // 标题
    UILabel *title = (UILabel *)[_contentView viewWithTag:801];
    if (title) title.textColor = textC;
    UILabel *name = (UILabel *)[_contentView viewWithTag:802];
    if (name) name.textColor = textC;

    // 夜间模式卡片
    UIView *themeCard = [_contentView viewWithTag:820];
    if (themeCard) themeCard.backgroundColor = cardC;
    UILabel *themeLbl = (UILabel *)[_contentView viewWithTag:821];
    if (themeLbl) themeLbl.textColor = textC;

    // 目标卡片
    UIView *goalCard = [_contentView viewWithTag:830];
    if (goalCard) goalCard.backgroundColor = cardC;
    UILabel *goalTitle = (UILabel *)[_contentView viewWithTag:831];
    if (goalTitle) goalTitle.textColor = textC;

    _themeSwitch.on = dark;
}

@end

File: health_advanced_swift/HealthDashboard/Controllers/HDProfileViewController.swift
//
//  HDProfileViewController.swift
//  HealthDashboard - 个人中心
//
//  Created by AI Assistant on 2026/3/22.
//

import UIKit
import Combine

// MARK: - ViewModel
// 简化版 ViewModel，只负责存储 UI 状态，不访问 Model

@MainActor
class HDProfileViewModel: ObservableObject {
    @Published var todayStepsText: String = "0"
    @Published var waterMLText: String = "0ml"
    @Published var sleepHoursText: String = "0h"
    
    @Published var stepsGoalText: String = "10000"
    @Published var waterGoalText: String = "2000"
    
    @Published var isDarkMode: Bool = false
    @Published var backgroundColor: UIColor = .white
    @Published var cardBackgroundColor: UIColor = .white
    @Published var textColor: UIColor = .black
    
    private var cancellables = Set<AnyCancellable>()
    
    init() {
        // 空初始化，数据由 ViewController 设置
    }
    
    func updateThemeColors() {
        let dark = isDarkMode
        backgroundColor = dark ? UIColor(red: 0.06, green: 0.08, blue: 0.14, alpha: 1.0)
                               : UIColor(red: 0.93, green: 0.95, blue: 0.97, alpha: 1.0)
        cardBackgroundColor = dark ? UIColor(red: 0.12, green: 0.16, blue: 0.24, alpha: 1.0)
                                   : UIColor.white
        textColor = dark ? UIColor.white
                        : UIColor(red: 0.10, green: 0.12, blue: 0.18, alpha: 1.0)
    }
}

// MARK: - ViewController

@MainActor
class HDProfileViewController: UIViewController {
    private let viewModel = HDProfileViewModel()
    private var cancellables = Set<AnyCancellable>()
    
    private let scrollView = UIScrollView()
    private let contentView = UIView()
    private let themeSwitch = UISwitch()
    
    // UI 标签引用
    private var titleLabel: UILabel?
    private var nameLabel: UILabel?
    private var themeCard: UIView?
    private var themeLbl: UILabel?
    private var goalCard: UIView?
    private var goalTitle: UILabel?
    private var statLabels: [UILabel] = []
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupScrollView()
        buildContent()
        setupBindings()
        applyTheme()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        navigationController?.isNavigationBarHidden = true
        applyTheme()
    }
    
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        scrollView.frame = view.bounds
    }
    
    // MARK: - UI Setup
    
    private func setupScrollView() {
        let sw = UIScreen.main.bounds.size.width
        let sh = UIScreen.main.bounds.size.height
        
        scrollView.frame = CGRect(x: 0, y: 0, width: sw, height: sh)
        scrollView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        scrollView.showsVerticalScrollIndicator = false
        view.addSubview(scrollView)
        
        contentView.frame = CGRect(x: 0, y: 0, width: sw, height: 600)
        scrollView.addSubview(contentView)
        scrollView.contentSize = CGSize(width: sw, height: 600)
    }
    
    private func buildContent() {
        let sw = view.bounds.size.width
        let topY: CGFloat = 60
        
        // 页面标题
        let title = UILabel(frame: CGRect(x: 20, y: topY, width: sw - 40, height: 40))
        title.text = "个人中心"
        title.font = UIFont.systemFont(ofSize: 28, weight: .heavy)
        title.textColor = .white
        contentView.addSubview(title)
        titleLabel = title
        
        // 头像
        let avatar = UIView(frame: CGRect(x: sw/2 - 44, y: topY + 58, width: 88, height: 88))
        avatar.backgroundColor = UIColor(red: 0.22, green: 0.54, blue: 0.95, alpha: 1)
        avatar.layer.cornerRadius = 44
        contentView.addSubview(avatar)
        
        let avatarEmoji = UILabel(frame: avatar.bounds)
        avatarEmoji.text = "😊"
        avatarEmoji.font = UIFont.systemFont(ofSize: 44)
        avatarEmoji.textAlignment = .center
        avatar.addSubview(avatarEmoji)
        
        // 用户名
        let name = UILabel(frame: CGRect(x: 20, y: topY + 158, width: sw - 40, height: 28))
        name.text = "健康达人"
        name.font = UIFont.systemFont(ofSize: 20, weight: .bold)
        name.textColor = .white
        name.textAlignment = .center
        contentView.addSubview(name)
        nameLabel = name
        
        // 统计行
        let statTitles = ["今日步数", "喝水", "睡眠"]
        let itemW = sw / 3.0
        let statsY = topY + 200
        
        for i in 0..<3 {
            let val = UILabel(frame: CGRect(x: CGFloat(i) * itemW, y: statsY, width: itemW, height: 32))
            val.font = UIFont.monospacedDigitSystemFont(ofSize: 20, weight: .bold)
            val.textColor = UIColor(red: 0.22, green: 0.54, blue: 0.95, alpha: 1)
            val.textAlignment = .center
            contentView.addSubview(val)
            statLabels.append(val)
            
            let lbl = UILabel(frame: CGRect(x: CGFloat(i) * itemW, y: statsY + 34, width: itemW, height: 18))
            lbl.text = statTitles[i]
            lbl.font = UIFont.systemFont(ofSize: 12)
            lbl.textColor = UIColor(white: 0.55, alpha: 1)
            lbl.textAlignment = .center
            contentView.addSubview(lbl)
        }
        
        // 夜间模式卡片
        let themeY = statsY + 72
        let themeCardView = UIView(frame: CGRect(x: 16, y: themeY, width: sw - 32, height: 64))
        themeCardView.backgroundColor = UIColor(red: 0.12, green: 0.16, blue: 0.24, alpha: 1)
        themeCardView.layer.cornerRadius = 16
        contentView.addSubview(themeCardView)
        themeCard = themeCardView
        
        let themeLblView = UILabel(frame: CGRect(x: 16, y: 18, width: sw - 100, height: 28))
        themeLblView.text = "🌙  夜间模式"
        themeLblView.font = UIFont.systemFont(ofSize: 16, weight: .semibold)
        themeLblView.textColor = .white
        themeCardView.addSubview(themeLblView)
        themeLbl = themeLblView
        
        themeSwitch.frame = CGRect(x: sw - 32 - 70, y: 16, width: 51, height: 31)
        themeSwitch.onTintColor = UIColor(red: 0.22, green: 0.54, blue: 0.95, alpha: 1)
        themeSwitch.addTarget(self, action: #selector(themeSwitchChanged), for: .valueChanged)
        themeCardView.addSubview(themeSwitch)
        
        // 目标卡片
        let goalY = themeY + 80
        let goalCardView = UIView(frame: CGRect(x: 16, y: goalY, width: sw - 32, height: 120))
        goalCardView.backgroundColor = UIColor(red: 0.12, green: 0.16, blue: 0.24, alpha: 1)
        goalCardView.layer.cornerRadius = 16
        contentView.addSubview(goalCardView)
        goalCard = goalCardView
        
        let goalTitleView = UILabel(frame: CGRect(x: 16, y: 14, width: sw - 64, height: 22))
        goalTitleView.text = "🎯  我的目标"
        goalTitleView.font = UIFont.systemFont(ofSize: 16, weight: .semibold)
        goalTitleView.textColor = .white
        goalCardView.addSubview(goalTitleView)
        goalTitle = goalTitleView
        
        // 目标内容（动态绑定）
        let goalLabel1 = UILabel(frame: CGRect(x: 16, y: 44, width: sw - 64, height: 24))
        goalLabel1.font = UIFont.systemFont(ofSize: 14)
        goalLabel1.textColor = UIColor(white: 0.70, alpha: 1)
        goalCardView.addSubview(goalLabel1)
        
        let goalLabel2 = UILabel(frame: CGRect(x: 16, y: 76, width: sw - 64, height: 24))
        goalLabel2.font = UIFont.systemFont(ofSize: 14)
        goalLabel2.textColor = UIColor(white: 0.70, alpha: 1)
        goalCardView.addSubview(goalLabel2)
        
        // 绑定目标标签
        viewModel.$stepsGoalText
            .map { "每日步数：\($0) 步" }
            .receive(on: DispatchQueue.main)
            .sink { [weak goalLabel1] text in
                goalLabel1?.text = text
            }
            .store(in: &cancellables)
        
        viewModel.$waterGoalText
            .map { "每日喝水：\($0) ml" }
            .receive(on: DispatchQueue.main)
            .sink { [weak goalLabel2] text in
                goalLabel2?.text = text
            }
            .store(in: &cancellables)
    }
    
    private func setupBindings() {
        // 从 Model 读取初始数据
        refreshData()

        // 绑定统计数据到 UI
        viewModel.$todayStepsText
            .receive(on: DispatchQueue.main)
            .sink { [weak self] text in self?.statLabels[safe: 0]?.text = text }
            .store(in: &cancellables)

        viewModel.$waterMLText
            .receive(on: DispatchQueue.main)
            .sink { [weak self] text in self?.statLabels[safe: 1]?.text = text }
            .store(in: &cancellables)

        viewModel.$sleepHoursText
            .receive(on: DispatchQueue.main)
            .sink { [weak self] text in self?.statLabels[safe: 2]?.text = text }
            .store(in: &cancellables)

        // 绑定主题开关状态
        viewModel.$isDarkMode
            .receive(on: DispatchQueue.main)
            .sink { [weak self] isDark in self?.themeSwitch.isOn = isDark }
            .store(in: &cancellables)

        // 订阅 Model @Published 属性替代通知
        Publishers.MergeMany(
            HDHealthDataModel.shared.$todaySteps.map { _ in () }.eraseToAnyPublisher(),
            HDHealthDataModel.shared.$waterML.map { _ in () }.eraseToAnyPublisher(),
            HDHealthDataModel.shared.$sleepHours.map { _ in () }.eraseToAnyPublisher()
        )
        .debounce(for: .milliseconds(50), scheduler: DispatchQueue.main)
        .sink { [weak self] in self?.refreshData() }
        .store(in: &cancellables)

        HDHealthDataModel.shared.$isDarkMode
            .receive(on: DispatchQueue.main)
            .sink { [weak self] _ in
                self?.refreshData()
                self?.applyTheme()
            }
            .store(in: &cancellables)

        refreshData()
    }

    private func refreshData() {
        let model = HDHealthDataModel.shared
        viewModel.todayStepsText = "\(model.todaySteps)"
        viewModel.waterMLText = String(format: "%.0fml", model.waterML)
        viewModel.sleepHoursText = String(format: "%.1fh", model.sleepHours.last ?? 0)
        viewModel.stepsGoalText = "\(model.stepsGoal)"
        viewModel.waterGoalText = String(format: "%.0f", model.waterGoalML)
        viewModel.isDarkMode = model.isDarkMode
        viewModel.updateThemeColors()
    }
    
    // MARK: - 主题切换
    
    @objc private func themeSwitchChanged() {
        HDHealthDataModel.shared.isDarkMode.toggle()
        applyTheme()
    }
    
    @objc func applyTheme() {
        view.backgroundColor = viewModel.backgroundColor
        scrollView.backgroundColor = viewModel.backgroundColor
        contentView.backgroundColor = viewModel.backgroundColor
        
        titleLabel?.textColor = viewModel.textColor
        nameLabel?.textColor = viewModel.textColor
        
        themeCard?.backgroundColor = viewModel.cardBackgroundColor
        themeLbl?.textColor = viewModel.textColor
        
        goalCard?.backgroundColor = viewModel.cardBackgroundColor
        goalTitle?.textColor = viewModel.textColor
    }
}

// MARK: - Array Safe Access Extension

extension Array {
    subscript(safe index: Int) -> Element? {
        return indices.contains(index) ? self[index] : nil
    }
}
